File: src/面试题08_08.permutation-ii-lcci.100348/solution.ts
function permutation (S: string): string[] {
    const sList = S.split('').sort();
    const used:boolean[] = new Array(sList.length).fill(false);
    const result:string[] = [];
    backTracking(sList, used, [] as string[], result);
    return result;
}

function backTracking (sList:string[], used:boolean[], sequence:string[], result:string[]) {
    if (sequence.length === sList.length) {
        result.push(sequence.join(''));
        return;
    }
    for (let i = 0; i < sList.length; i++) {
        if (used[i] || (i > 0 && sList[i - 1] === sList[i] && !used[i - 1])) {
            continue;
        }
        used[i] = true;
        sequence.push(sList[i]);
        backTracking(sList, used, sequence, result);
        sequence.pop();
        used[i] = false;
    }
}

File: src/面试题08_08.permutation-ii-lcci.100348/solution.java
class Solution {
    char[] chars;
    boolean[] used;
    StringBuilder stringBuilder = new StringBuilder();
    List<String> result = new ArrayList<>();
    public String[] permutation(String S) {
        chars = S.toCharArray();
        Arrays.sort(chars);
        used = new boolean[chars.length];

        backTracking();

        return result.toArray(new String[0]);
    }

    private void backTracking(){
        if(stringBuilder.length() == chars.length){
            result.add(stringBuilder.toString());
            return;
        }
        for(int i=0;i< chars.length; i++){
            if(used[i] || (i>0 && chars[i] == chars[i-1] && !used[i-1])){
                continue;
            }
            used[i] = true;
            stringBuilder.append(chars[i]);
            backTracking();
            stringBuilder.deleteCharAt(stringBuilder.length()-1);
            used[i] = false;
        }

    }
}